### PR TITLE
Add support for dub

### PR DESCRIPTION
**Pull request by John-Colvin**
http://code.dlang.org is becoming the de facto standard as a D package repository, along with it's associated build tool `dub`. Adding this package.json file and then publishing to the code.dlang repo\* will allow people to express tango as a dependency for their project and have `dub` deal with downloading from this github repo and building.

There's a lot of quality code in tango, it seems sensible that it should be available through `dub`.

*I will do this if/when this pull is merged.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{
+    "name": "tango-d2",
```
Comment (jacob-carlborg) [inline package.json]: Why not just "tango"?

File: package.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    "description": "A large general-purpose library. A port of the D1 Tango library.",
+    "homepage": "https://github.com/SiegeLord/Tango-D2/",
```
Comment (jacob-carlborg) [inline package.json:4]: I think there should be a link to the original site, somewhere.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    "license": "BSD",
+    "authors": [
+        "SiegeLord",
+        "Origin Tango authors"
```
Comment (jacob-carlborg) [inline package.json]: Shouldn't the original authors be listed as well?

**Comment by jacob-carlborg** [comment]
You need to add the following versions: "darwin" and "osx" for Mac OS X. "freebsd" for FreeBSD.

File: package.json (hunk 3 above)
Comment (John-Colvin) [inline package.json]: all of them? surely that's a very large number of people.

File: package.json (hunk 1 above)
Comment (John-Colvin) [inline package.json]: fixed

File: package.json (hunk 2 above)
Comment (John-Colvin) [inline package.json:4]: I would say that's the job of the readme here, not the package.json

File: package.json (hunk 3 above)
Comment (jacob-carlborg) [inline package.json]: I was thinking Kris Bell, Sean Kelly and Lars Ivar Igesund.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{
+    "name": "tango",
+    "description": "A large general-purpose library. A port of the D1 Tango library.",
+    "homepage": "https://github.com/SiegeLord/Tango-D2/",
+    "license": "BSD",
```
Comment (jacob-carlborg) [inline package.json]: According to http://code.dlang.org/package-format#licenses, this should be "BSD 3-clause".

File: package.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        "Origin Tango authors"
+    ],
+
+    "sourcePaths": ["tango/core", "tango/io", "tango/math", "tango/net", 
+		    "tango/stdc", "tango/text", "tango/time", "tango/util"],
+    "sourcePaths-Windows": ["tango/sys/win32"],
+    "sourcePaths-linux": ["tango/sys/linux"],
+    "sourcePaths-FreeBSD": ["tango/sys/freebsd"],
+    "sourcePaths-Solaris": ["tango/sys/solaris"],
+    "sourcePaths-OSX": ["tango/sys/darwin"],
```
Comment (jacob-carlborg) [inline package.json]: tango/stdc/posix needs to be excluded on Windows.

Comment (jacob-carlborg) [inline package.json]: I don't know if the version names are case insensitive in package.json but all examples here: http://code.dlang.org/package-format, use lower case.

**Comment by SiegeLord** [comment]
Is it onerous to put the package.json into the build directory (not a big deal if it is)?

Can a static library configuration be added?

Should this be fixed:

```
Package tango (configuration "shared") defines no import paths, use {"importPaths": [...]} or the default package directory structure to fix this.
```

File: package.json (hunk 5 above)
Comment (John-Colvin) [inline package.json]: I don't think stdc/posix is excluded on windows in the makefile, which is what I was basing this on. However, I see your pull to enable glob matching in dub has been merged, so I've gone ahead and used that.

**Comment by John-Colvin** [comment]
> Is it onerous to put the package.json into the build directory (not a big deal if it is)?

Unfortunately it has to be in the root folder, as far as I know.

> Can a static library configuration be added?

Done

> Should this be fixed:
> Package tango (configuration "shared") defines no import paths...

Done

**Comment by John-Colvin** [comment]
> I don't know if the version names are case insensitive in package.json but all examples here: http://code.dlang.org/package-format, use lower case.

I'm not sure on this. It refers to the normal dlang version specifiers which are mostly capitalised, but then immediately uses them as lower-case in the example

File: package.json (hunk 4 above)
Comment (John-Colvin) [inline package.json]: fixed

File: package.json (hunk 3 above)
Comment (John-Colvin) [inline package.json]: D1/tango is before my time so I'll take your word for it and add them explicitly

File: package.json (hunk 5 above)
Comment (jacob-carlborg) [inline package.json]: Both the Ruby script `bob.rb` and the `bob` binary excludes "tango/stdc/posix" on Windows. These are part of the original build scripts. The makefile is new in the D2 port. Not all declarations are contained in version blocks. Thanks.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    "excludedSourceFiles-Windows": ["tango/stdc/posix/*"],
+
+    "importPaths": [""],
+
+    "configurations": [
+	{
+            "name": "static",
+            "targetType": "staticLibrary"
```
Comment (jacob-carlborg) [inline package.json:32]: Should this really be a configuration. Shouldn't this be at the top level to be default?

Comment (John-Colvin) [inline package.json:32]: It appears that the first configuration is used by default, so in order to have static as the default this is what has to be done. Without the static config the shared config is always used.

Comment (jacob-carlborg) [inline package.json:32]: Ok, I see.

**Comment by jacob-carlborg** [comment]
Any update on this?

**Comment by John-Colvin** [comment]
I got confirmation on the capitalisation: lowercase is the correct format. Fixed.

**Comment by jacob-carlborg** [comment]
What about the custom version identifiers that need to be added: https://github.com/SiegeLord/Tango-D2/pull/57#issuecomment-27653518.

**Comment by John-Colvin** [comment]
Sorry, I wasn't aware tango used non-standard version identifiers in places. A quick `grep -roh 'version(.*)' . | sort | uniq` in the tango folder shows up a few: darwin*, openbsd and solaris. If I include the unported directory then there is freebsd as well. I also noticed PPC_CPU and X86_CPU. I couldn't find osx anywhere so I guess it's not needed anymore.

*is standard, but deprecated.

Why are these here? Shouldn't they be swapped out for their standard counterparts?

**Comment by jacob-carlborg** [comment]
> Sorry, I wasn't aware tango used non-standard version identifiers in places. A quick `grep -roh 'version(.*)' . | sort | uniq` in the tango folder shows up a few: darwin*, openbsd and solaris. If I include the unported directory then there is freebsd as well. I also noticed PPC_CPU and X86_CPU. I couldn't find osx anywhere so I guess it's not needed anymore.

The ones I listed is what's used by the Ruby and D build scripts.

> Why are these here? Shouldn't they be swapped out for their standard counterparts?

They were standard, until Walter ported DMD to Mac OS X and FreeBSD and decided to change well established version identifiers.

I guess they could be replaced with the now standard counterparts. But this probably should be in another pull request. I would really _not_ like this pull request to be put on hold because of that.

**Comment by SiegeLord** [comment]
So can these versions be specified in the dub file? Or is that just an impossibility?

**Comment by jacob-carlborg** [comment]
Yes, they can.

**Comment by John-Colvin** [comment]
Which ones should I specify? As @jacob-carlborg says, "darwin", "osx" and "freebsd" are used in bob, but this is not completely congruent with what can be found in the code. E.g. "freebsd" is only present in unported, "osx" is entirely absent, "solaris" and "openbsd" are present in the code but not in bob etc...

I suggest to only specify what bob does, for consistency between build tools. Then even if it's wrong/lacking, it won't be missed due to individual tango developers using one tool or the other. Then if/when there is some standardisation and a careful review of version blocks by someone familiar with the code, both tools can be corrected appropriately. Does that sound reasonable?

**Comment by John-Colvin** [comment]
I've done the above for now, perhaps we are good to go?

**Comment by jacob-carlborg** [comment]
LGTM :ok_hand:

**Comment by John-Colvin** [comment]
I am happy to manage the registering on code.dlang.org and maintain it for now, unless someone else particularly wants to. If not then I'll register it when this pull is merged.

**Comment by SiegeLord** [comment]
Ok. Do you mind rebasing so I don't get that pointless merge commit? Thanks.

**Comment by John-Colvin** [comment]
is that better? I'm not great at git but I think I got it right.

**Comment by SiegeLord** [comment]
Thanks a lot!

**Comment by John-Colvin** [comment]
just hit a blockage with registering as code.dlang.org expects to work with the master branch (https://github.com/rejectedsoftware/dub-registry/issues/31). Hopefully this restriction will be lifted quickly (it's perfectly reasonable to want to have a dedicated dub release branch, for example), but in the mean time:

Is there a reason why the master branch is effectively a old, dead version of d2port? Normally one would expect the master branch to be the main point of reference for the state of a project. Would it be possible/desirable/sensible to just completely overwrite it from d2port and switch it to be the default?

**Comment by jacob-carlborg** [comment]
It's also possible to use a tag.

**Comment by John-Colvin** [comment]
Thanks to quick work from the dub guys, we're live: http://code.dlang.org/packages/tango

:)

**Comment by jacob-carlborg** [comment]
Awesome :smiley:

**Pull request merged**